common/flatpak-dir.c

```diff
@@ -7959,11 +7959,29 @@ flatpak_dir_user_check_parental_controls (FlatpakDir    *self,
 {
   g_autoptr(AsApp) app = NULL;
   g_autoptr(GError) local_error = NULL;
+  const char *on_session = g_getenv ("FLATPAK_SYSTEM_HELPER_ON_SESSION");
+  g_autoptr(GDBusConnection) dbus_connection = NULL;
 
   /* The system helper implements its own checks via polkit. */
   if (!self->user)
     return TRUE;
 
+  dbus_connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, cancellable, &local_error);
+  if (dbus_connection == NULL)
+    {
+      if (on_session != NULL)
+        {
+          g_debug ("Skipping parental controls check for %s since the "
+                   "system bus is unavailable in the test environment", ref);
```

Comment:
Doesn't this allow a user to bypass parental controls by running:

```
FLATPAK_SYSTEM_HELPER_ON_SESSION=1 DBUS_SYSTEM_BUS_ADDRESS=/dev/null flatpak install --user flathub very.violent.game
```

Reply:
Ah, I thought about that possibility but didn't know about `DBUS_SYSTEM_BUS_ADDRESS`

Reply:
The cheap way to work around that would be to check if `DBUS_SYSTEM_BUS_ADDRESS` is set when deciding whether to propagate the error, but there are probably legitimate reasons to set that

Reply:
Maybe we can copy what's done in the tests in `eos-parental-controls` but it will require a dependency on `libglib-testing`

Reply:
Upon further investigation I don't know of a reasonable way to do this. If `testlibrary` creates the `GDBusConnection` based on a `GtDBusQueue` there's no reasonable way for it to pass that to libflatpak where it's needed. And if libflatpak creates the `GDBusConnection` we would still need a way to distinguish the test environment from production, like a `FlatpakInstallFlags` value but that seems messy.

I wonder if we could just assume kids won't know to set these environment variables? Couldn't a determined user also just compile a version of flatpak that doesn't do parental controls checks?

Reply:
You either have to provide a way to disable parental controls checks, or mock up the parental controls D-Bus service (which is actually part of `org.freedesktop.Accounts`). While I’ve got a suggestion for how to do the latter, it’s a few hours’ work so might make sense to leave until after the rebase. If that’s the case, then I can’t see anything immediately better than your current patch (plus a `FIXME` comment, plus a follow-up issue).

My longer-term suggestion:
 1. Drop `FLATPAK_SYSTEM_HELPER_ON_SESSION` and instead set `DBUS_SYSTEM_BUS_ADDRESS`, just to simplify the code a bit. You can get the address to set it to from `g_test_dbus_get_bus_address()` in `testlibrary.c`.
 2. Add a service file in `tests/services` (`services_dir` in `testlibrary.c`) which advertises `org.freedesktop.Accounts` and has its `Exec=` line as a simple Python program which mocks that service just enough to allow parental controls to pass. That would mean listing one or two dummy accounts, and returning the default parental controls properties for them. It would be about 100 lines, I’m guessing, and wouldn’t need to change its behaviour for different flatpak tests.

If we ever wanted to add unit tests for parental controls interaction between the system helper and accounts-daemon, that’s a bigger problem which is out of scope here. It might involve `GtDBusQueue`, although as the amount of code under test gets bigger, `GtDBusQueue` becomes less appropriate (as you end up setting up a huge amount of state for each test). `python-dbusmock` might be more appropriate, to implement a mock accounts-daemon service whose behaviour can be tweaked by other D-Bus methods it exposes.

Reply:
Added the FIXME comment